api/dimensions.go

```diff
@@ -139,41 +157,53 @@ func (api *DatasetAPI) getDimensionOptions(w http.ResponseWriter, r *http.Reques
 		state = models.PublishedState
 	}
 
-	b, err := func() ([]byte, error) {
-		version, err := api.dataStore.Backend.GetVersion(datasetID, edition, versionID, state)
-		if err != nil {
-			log.Event(ctx, "failed to get version", log.ERROR, log.Error(err), logData)
-			return nil, err
-		}
+	// get limit from query parameters, or default value
+	limit, err := getPositiveIntQueryParameter(r.URL.Query(), "limit", api.defaultLimit)
+	if err != nil {
+		logData["query_params"] = r.URL.RawQuery
+		handleDimensionsErr(ctx, w, "failed to obtain limit from request query paramters", err, logData)
+		return
+	}
 
-		if err = models.CheckState("version", version.State); err != nil {
-			logData["version_state"] = version.State
-			log.Event(ctx, "unpublished version has an invalid state", log.ERROR, log.Error(err), logData)
-			return nil, err
-		}
+	// get offset from query parameters, or default value
+	offset, err := getPositiveIntQueryParameter(r.URL.Query(), "offset", api.defaultOffset)
+	if err != nil {
+		logData["query_params"] = r.URL.RawQuery
+		handleDimensionsErr(ctx, w, "failed to obtain offset from request query paramters", err, logData)
+		return
+	}
 
-		results, err := api.dataStore.Backend.GetDimensionOptions(version, dimension)
-		if err != nil {
-			log.Event(ctx, "failed to get a list of dimension options", log.ERROR, log.Error(err), logData)
-			return nil, err
-		}
+	// ger version for provided dataset, edition and versionID
+	version, err := api.dataStore.Backend.GetVersion(datasetID, edition, versionID, state)
+	if err != nil {
+		handleDimensionsErr(ctx, w, "failed to get version", err, logData)
+		return
+	}
 
-		for i := range results.Items {
-			results.Items[i].Links.Version.HRef = fmt.Sprintf("%s/datasets/%s/editions/%s/versions/%s",
-				api.host, datasetID, edition, versionID)
-			results.Items[i].Links.Version.ID = versionID
-		}
+	// vaidate state
+	if err = models.CheckState("version", version.State); err != nil {
+		logData["version_state"] = version.State
+		handleDimensionsErr(ctx, w, "unpublished version has an invalid state", err, logData)
+		return
+	}
 
-		b, err := json.Marshal(results)
-		if err != nil {
-			log.Event(ctx, "failed to marshal list of dimension option resources into bytes", log.ERROR, log.Error(err), logData)
-			return nil, err
-		}
+	// get sorted dimension options, starting at offset index, with a limit on the number of items
+	results, err := api.dataStore.Backend.GetDimensionOptions(version, dimension, offset, limit)
+	if err != nil {
+		handleDimensionsErr(ctx, w, "failed to get a list of dimension options", err, logData)
+		return
+	}
 
-		return b, nil
-	}()
+	// populate links
+	for i := range results.Items {
+		results.Items[i].Links.Version.HRef = fmt.Sprintf("%s/datasets/%s/editions/%s/versions/%s",
```

Comment:
It would be more efficient to create a var for the uri format outside the loop and reuse it.

```
versionFmt := "%s/datasets/%s/editions/%s/versions/%s"
for i := range results.Items {
		results.Items[i].Links.Version.HRef = fmt.Sprintf(versionFmt, api.host, datasetID, edition, versionID)
                 ...
}
```

Reply:
And actually, it's always the same value